README.md
# ETF2L
[![Build Status](https://travis-ci.org/nTraum/ETF2L.svg?branch=master)](https://travis-ci.org/nTraum/ETF2L)
[![Coverage Status](https://img.shields.io/coveralls/nTraum/ETF2L.svg)](https://coveralls.io/r/nTraum/ETF2L)
[![Dependency Status](https://gemnasium.com/nTraum/ETF2L.svg)](https://gemnasium.com/nTraum/ETF2L)
[![Code Climate](https://codeclimate.com/github/nTraum/ETF2L.png)](https://codeclimate.com/github/nTraum/ETF2L)

Ruby wrapper for the [ETF2L API](http://api.etf2l.org/).

## Installation

Add this line to your application's Gemfile:

```ruby
gem "etf2l"
```

And then execute:

```sh
$ bundle
```

Or install it yourself as:

```sh
$ gem install etf2l
```

## Usage

### Whitelists

Fetch an array of whitelists currently used in the ETF2L:

```ruby
whitelists = Etf2l::API.whitelists
whitelists.first.name #=> "ETF2L Ultiduo"
whitelists.first.url  #=> "http://etf2l.org/configs/etf2l_whitelist_ultiduo.txt"
```

## Contributing

1. Fork it ( https://github.com/nTraum/ETF2L/fork )
2. Create your feature branch (`git checkout -b my-new-feature`)
3. Commit your changes (`git commit -am 'Add some feature'`)
4. Push to the branch (`git push origin my-new-feature`)
5. Create a new Pull Request


CHANGELOG.md
# 0.1 (unreleased)
 * Initial release


spec/spec_helper.rb
require "simplecov"
require "coveralls"

SimpleCov.formatter = SimpleCov::Formatter::MultiFormatter[
  Coveralls::SimpleCov::Formatter,
  SimpleCov::Formatter::HTMLFormatter
]
SimpleCov.start do
  add_filter "spec"
end

require "pry"
require "vcr"

require "etf2l"
include Etf2l

VCR.configure do |c|
  c.cassette_library_dir = "spec/cassettes"
  c.hook_into :webmock
  c.configure_rspec_metadata!
end

RSpec.configure do |config|
  if config.files_to_run.one?
    config.default_formatter = "doc"
  end

  config.profile_examples = 10
  config.order = :random
  Kernel.srand config.seed

  config.expect_with :rspec do |expectations|
    expectations.syntax = :expect
  end

  config.mock_with :rspec do |mocks|
    mocks.syntax = :expect
    mocks.verify_partial_doubles = true
  end
end


spec/lib/etf2l/api_spec.rb
describe API, vcr: true do
  context "when the player is fetched by id" do
    let(:identifier) { 10512 }
    subject(:player) { described_class.player(identifier) }

    it "fetches the bans" do
      expect(player.bans).to be_empty
      expect(player.banned?).to eq(false)
    end

    it "fetches the classes" do
      expect(player.classes).to include("Demoman", "Soldier")
    end

    it "fetches the country" do
      expect(player.country).to eq("Germany")
    end

    it "fetches the id" do
      expect(player.id).to eq(identifier)
    end

    it "fetches the name" do
      expect(player.name).to eq("nTraum")
    end

    it "fetches the nickname" do
      expect(player.nickname).to eq("nTraum")
    end

    it "fetches the title" do
      expect(player.title).to eq("Site Admin")
    end

    it "fetches the steam avatar" do
      expect(player.steam.avatar).to eq("http://cdn.akamai.steamstatic.com/steamcommunity/public/images/avatars/65/65044ba608c1a1c03a163e9d8291871e2e9474cf_full.jpg")
    end

    it "fetches the SteamID" do
      expect(player.steam.id).to eq("STEAM_0:0:18721806")
    end

    it "fetches the SteamID64" do
      expect(player.steam.id64).to eq(76561197997709340)
    end
  end

  context "when the player is currently banned" do
    let(:identifier) { 91596 }
    subject(:player) { described_class.player(identifier) }

    it "is banned" do
      expect(player.banned?).to eq(true)
    end

    it "fetches the ban start time" do
      expect(player.bans.first.start).to eq(Time.at(1398463200))
    end

    it "fetches the ban end time" do
      expect(player.bans.first.end).to eq(Time.at(1429999200))
    end

    it "fetches the ban reason" do
      expect(player.bans.first.reason).to eq("VAC Ban")
    end
  end

  it "fetches the whitelists" do
    whitelists = API.whitelists
    expect(whitelists).to be_an(Array)
    expect(whitelists.first).to be_a(Whitelist)
    expect(whitelists.first.name).to eq("ETF2L Ultiduo")
  end
end


spec/lib/etf2l/ban_spec.rb
describe Ban do
  it "calculates the duration of the ban" do
    ban = Ban.new({ "start" => 200, "end" => 300 })

    expect(ban.duration).to eq(100)
  end
end


lib/etf2l/resources/player.rb
require "memoizable"

module Etf2l

  # Player of the ETF2L.
  class Player

    include Memoizable

    # @!attribute [r] bans
    #   @return [Array<Ban>] An array of bans currently affecting the player.
    # @!attribute [r] country
    #   @return [String] nationality of the player.
    attr_reader :bans,
                :country,
                :classes,
                :id,
                :name,
                :title,
                :steam

    alias :nickname :name
    alias :nationality :country

    def initialize(attributes)
      @bans = []
      @country = attributes["country"]
      @classes = attributes["classes"]
      @id = attributes["id"]
      @name = attributes["name"]
      @title = attributes["title"]
      @steam = Steam.new(attributes["steam"])

      if attributes["bans"]
        attributes["bans"].each { |ban| @bans << Ban.new(ban) }
      end
    end

    # Returns true if the player is currently banned.
    # @return [Boolean] true the player is currently banned.
    def banned?
      bans.any?
    end

    memoize :banned?
  end
end


lib/etf2l/api.rb
require "httparty"

module Etf2l
  class API
    include HTTParty
    base_uri "http://api-v1.etf2l.org"
    headers "accept" => "application/json"
    format :json

    # Returns the requested player.
    # @param identifier [Fixnum, String] ETF2L ID, SteamID or SteamID64 of
    #   the player.
    #
    # @return [Player] The requested player.
    # @example by ETF2L ID
    #   API.player(10512)
    # @example by SteamID
    #   API.player("STEAM_0:0:18721806")
    # @example by SteamID64
    #   API.player(76561197997709340)
    def self.player(identifier)
      response = get("/player/#{identifier}")
      Player.new(response["player"])
    end


    # Returns an array of (currently used) whitelists.
    # @example
    #  API.whitelists
    # @return [Array<Whitelist>] an array of (currently used) whitelists.
    def self.whitelists
      response = get("/whitelists")
      response["whitelists"].map do |name, attributes|
        Whitelist.new(attributes.merge("name" => name))
      end
    end
  end
end
